File: labs/lab1_ws/src/lab1_pkg/src/talker.cpp
#include <chrono>
#include <memory>
#include <rclcpp/rclcpp.hpp>
#include <ackermann_msgs/msg/ackermann_drive_stamped.hpp>

using namespace std::chrono_literals;

class Talker : public rclcpp::Node
{
public:
  Talker() : Node("talker")
  {
    // Declare parameters
    this->declare_parameter("v", 1.0);
    this->declare_parameter("d", 0.0);

    // Create publisher
    drive_pub_ = this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>("/drive", 10);

    // Create timer
    timer_ = this->create_wall_timer(100ms, std::bind(&Talker::publish_drive, this));
  }

private:
  void publish_drive()
  {
    // Get parameters
    double v = this->get_parameter("v").as_double();
    double d = this->get_parameter("d").as_double();

    // Create message
    auto msg = ackermann_msgs::msg::AckermannDriveStamped();
    msg.drive.speed = v;
    msg.drive.steering_angle = d;

    // Publish message
    drive_pub_->publish(msg);
    RCLCPP_INFO(this->get_logger(), "Publishing: speed=%.2f, steering=%.2f", v, d);
  }

  rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr drive_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<Talker>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
} 

File: labs/lab1_ws/scripts/talker.py
# lab1_pkg/talker.py
import sys
sys.path.append('/home/kipp/ackermann_ws/install/ackermann_msgs/lib/python3.8/site-packages')
import rclpy
from rclpy.node import Node
from ackermann_msgs.msg import AckermannDriveStamped

class Talker(Node):
    def __init__(self):
        super().__init__('talker')
        self.declare_parameter('v', 1.0)
        self.declare_parameter('d', 0.0)

        self.drive_pub = self.create_publisher(AckermannDriveStamped, '/drive', 10)
        self.timer = self.create_timer(0.1, self.publish_drive)

    def publish_drive(self):
        v = self.get_parameter('v').get_parameter_value().double_value
        d = self.get_parameter('d').get_parameter_value().double_value

        msg = AckermannDriveStamped()
        msg.drive.speed = v
        msg.drive.steering_angle = d
        self.drive_pub.publish(msg)
        self.get_logger().info(f'Publishing: speed={v:.2f}, steering={d:.2f}')

def main(args=None):
    rclpy.init(args=args)
    node = Talker()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()
